Sources/CodeEditSourceEditor/Controller/TextViewController+IndentLines.swift

```diff
@@ -0,0 +1,131 @@
+               let endLineInfo = textView.layoutManager.textLineForOffset(cursorPosition.range.upperBound),
+               startLineInfo.index != endLineInfo.index {
+                return true
+            }
+        }
+        return false
+    }
+
+    private func getHighlightedLines(for range: NSRange) -> [Int]? {
+        guard let startLineInfo = textView.layoutManager.textLineForOffset(range.lowerBound) else {
+            return nil
+        }
+        var lines: [Int] = [startLineInfo.index]
+
+        guard let endLineInfo = textView.layoutManager.textLineForOffset(range.upperBound),
+              endLineInfo.index != startLineInfo.index else {
+            return lines
+        }
+        if endLineInfo.index == startLineInfo.index + 1 {
+            lines.append(endLineInfo.index)
+            return lines
+        }
+
+        return Array(startLineInfo.index...endLineInfo.index)
+    }
+
+    private func adjustIndentation(lineIndex: Int, indentationChars: String, inwards: Bool) {
+        guard let lineInfo = textView.layoutManager.textLineForIndex(lineIndex) else { return }
+
+        if inwards {
+            if indentOption != .tab {
+                removeLeadingSpaces(lineInfo: lineInfo, spaceCount: indentationChars.count)
+            } else {
+                removeLeadingTab(lineInfo: lineInfo)
+            }
+        } else {
+            addIndentation(lineInfo: lineInfo, indentationChars: indentationChars)
+        }
+    }
+
+    private func addIndentation(
+        lineInfo: TextLineStorage<TextLine>.TextLinePosition,
+        indentationChars: String
+    ) {
+        textView.replaceCharacters(
+            in: NSRange(location: lineInfo.range.lowerBound, length: 0),
+            with: indentationChars
+        )
+    }
+
+    private func removeLeadingSpaces(
+        lineInfo: TextLineStorage<TextLine>.TextLinePosition,
+        spaceCount: Int
+    ) {
+        guard let lineContent = textView.textStorage.substring(from: lineInfo.range) else { return }
+
+        let removeSpacesCount = countLeadingSpacesUpTo(line: lineContent, maxCount: spaceCount)
+
+        guard removeSpacesCount > 0 else { return }
+
+        textView.replaceCharacters(
+            in: NSRange(location: lineInfo.range.lowerBound, length: removeSpacesCount),
+            with: ""
+        )
+    }
+
+    private func removeLeadingTab(lineInfo: TextLineStorage<TextLine>.TextLinePosition) {
+        guard let lineContent = textView.textStorage.substring(from: lineInfo.range) else {
+            return
+        }
+
+        if lineContent.first == "\t" {
+            textView.replaceCharacters(
+                in: NSRange(location: lineInfo.range.lowerBound, length: 1),
+                with: ""
+            )
+        }
+    }
+
+    private func countLeadingSpacesUpTo(line: String, maxCount: Int) -> Int {
+        // Count leading spaces using prefix and `filter`
+        return line.prefix(maxCount).filter { $0 == " " }.count
```

Comment:
This is slightly wrong, if I have a line like `/// ` and maxCount is `4` this returns `1` for the leading space count.

Reply:
I updated the function to use the old logic, i.e. a simple for loop which isn't as clean but is easy to understand and should work as expected